rustls/src/conn.rs

```diff
@@ -189,6 +189,20 @@ mod connection {
                 (false, false) => Err(io::ErrorKind::WouldBlock.into()),
             }
         }
+
+        /// Obtain a chunk of plaintext data received from the peer over this TLS connection.
+        /// This does the same thing as [`Reader::fill_buf`], but unlike that method, this
+        /// method returns a slice whose lifetime is bounded by the [`ConnectionCommon`]
+        /// that created this `Reader`.
+        pub fn into_first_chunk(self) -> io::Result<&'a [u8]> {
```

Comment:
Is there a reason this has to take `self` owned, instead of taking `&self`?

I don't think it makes sense to document this by pointing to `fill_buf()`, which doesn't exist yet and also feels like it creates a sort of circular reference. Just call out instead that the return value's lifetime extends to that of the underlying `ConnectionCommon`.

Reply:
Yes, because the inner reference is mutable. `&'a &'b mut T` can't convert to a `&'b T`, only `&'a T` (the mutable reference is only "temporarily" frozen for `'a`). This is unlike `&'a &'b T` where the inner `&'b T` can be copied out. Taking `self` by value instead consumes the `&'b mut T` borrow and lets us convert it into an immutable one.

The method could be written as taking `&'a self` but it would not be callable since the Reader is generally short lived and wouldn't be borrowable for `'a`.